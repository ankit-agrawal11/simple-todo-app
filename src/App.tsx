import * as React from 'react';
import { CheckBox } from 'react-native-elements';
import {
  StyleSheet,
  View,
  TextInput,
  Keyboard,
  Platform,
  AsyncStorage,
  FlatList,
  Button
} from "react-native";

const isAndroid = Platform.OS == "android";
const viewPadding = 10;

type ITaskProps = Array<{ key: string, text: string, isCompleted?: boolean }>;

export interface ITodoListState {
  tasks: ITaskProps;
  text: string;
  viewPadding?: string | number;
}

export default class App extends React.Component<{}, ITodoListState> {

  state = {
    tasks: [],
    text: "",
    viewPadding: viewPadding
  }

  async componentDidMount() {
    Keyboard.addListener(
      isAndroid ? "keyboardDidShow" : "keyboardWillShow",
      e => this.setState({ viewPadding: e.endCoordinates.height + viewPadding })
    );

    Keyboard.addListener(
      isAndroid ? "keyboardDidHide" : "keyboardWillHide",
      () => this.setState({ viewPadding: viewPadding })
    );

    const tasks = await this.getTasksFromStorage()
    this.setState({tasks});
  }

  addTask = () => {
    if (this.state.text.trim().length > 0) {
      this.setState(
        (prevState: ITodoListState) => {
          let { tasks, text } = prevState;
          const currentTaskIndex = tasks.length ? Number(tasks[tasks.length - 1].key) + 1 : 0;
          return {
            tasks: tasks.concat({ key: String(currentTaskIndex), text: text }),
            text: ""
          };
        },
        () => AsyncStorage.setItem("TASKS", JSON.stringify(this.state.tasks))
      );
    }
  };

  deleteTask = (index: number) => {
    this.setState(
      (prevState: ITodoListState) => {
        let tasks = prevState.tasks.slice();

        tasks.splice(index, 1);

        return { tasks: tasks };
      },
      () => AsyncStorage.setItem("TASKS", JSON.stringify(this.state.tasks))
    );
  };

  toggleStatus = (index: number) => {
    this.setState(
      (prevState: ITodoListState) => {
        let updatedTasks = prevState.tasks.map((task, taskIndex) => {
          if (taskIndex === index) {
            return { ...task, isCompleted: !task.isCompleted }
          }

          return task;
        });

        return { tasks: updatedTasks };
      }
    )
  }

  async getTasksFromStorage() {
    const items = await AsyncStorage.getItem("TASKS");

    return items && items.length ? JSON.parse(items) : [];
  }

  render(): JSX.Element {
    return (
      <View style={[styles.container, { paddingBottom: this.state.viewPadding }]}>
        <FlatList
          testID='flat-list'
          style={styles.list}
          data={this.state.tasks}
          renderItem={(task: { index: number, item: { text: string, key: string, isCompleted: boolean } }) =>
            <View>
              <View style={[styles.listItemCont]}>
                <View testID='checkbox' style={{ flex: 0.90 }}>
                  <CheckBox
                    textStyle={task.item.isCompleted ? {textDecorationLine: 'line-through'} : {}}
                    title={task.item.text}
                    checked={task.item.isCompleted}
                    onPress={() => this.toggleStatus(task.index)}
                  />
                </View>
                <View testID="delete-task-wrapper" style={{ flex: 0.10, alignItems: 'center', justifyContent: 'flex-end', marginRight: 1 }}>
                  <Button testID="delete-task" title="X" onPress={() => this.deleteTask(task.index)} />
                </View>
              </View>
              <View style={styles.hr} />
            </View>}
        />
        <TextInput
          testID='text-input'
          underlineColorAndroid='rgba(0,0,0,0)'
          style={styles.textInput}
          onChangeText={(text) => this.setState({ text: text })}
          onSubmitEditing={this.addTask}
          value={this.state.text}
          placeholder="Add Tasks"
          returnKeyType="done"
          returnKeyLabel="done"
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: "#F5FCFF",
    padding: viewPadding,
    paddingTop: 20
  },
  textInput: {
    height: 40,
    paddingRight: 10,
    paddingLeft: 10,
    borderColor: "gray",
    borderWidth: 1,
    width: "100%",
  },
  list: {
    width: "100%"
  },
  listItem: {
    paddingTop: 2,
    paddingBottom: 2,
    fontSize: 18
  },
  hr: {
    height: 1,
    backgroundColor: "gray",
    marginTop: 5,
    marginBottom: 5
  },
  listItemCont: {
    flexDirection: "row",
    alignItems: "center",
    flex: 1,
    justifyContent: "space-between"
  },
});
