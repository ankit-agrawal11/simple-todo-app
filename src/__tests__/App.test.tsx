import * as React from 'react';
import * as Renderer from 'react-test-renderer';

import App from '../App';

describe('Tests for App', () => {
  it('should match the snapshot.', () => {
    const snapshot = Renderer.create(<App />).toJSON();
    expect(snapshot).toMatchSnapshot();
  });
});
