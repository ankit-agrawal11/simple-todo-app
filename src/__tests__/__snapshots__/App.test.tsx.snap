// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for App should match the snapshot. 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#F5FCFF",
        "flex": 1,
        "justifyContent": "center",
        "padding": 10,
        "paddingTop": 20,
      },
      Object {
        "paddingBottom": 10,
      },
    ]
  }
>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "width": "100%",
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Add Tasks"
    returnKeyLabel="done"
    returnKeyType="done"
    style={
      Object {
        "borderColor": "gray",
        "borderWidth": 1,
        "height": 40,
        "paddingLeft": 10,
        "paddingRight": 10,
        "width": "100%",
      }
    }
    underlineColorAndroid="rgba(0,0,0,0)"
    value=""
  />
</View>
`;
